### ISSUE TITLE
linux: less kernel calls

### ISSUE THREAD

**Issue by obastemur** [Microsoft/ChakraCore#3930]
Made a quick experiment with locking algorithms and other cross thread sync implementations under PAL, saw 2% AcmeAIR LTO gain after basically updating from Sleep/Wait to PThread/Mutex

This tells me that  sometimes we lock longer than Sleep timer hence we end up Vol. context switching twice or more!

This might be causing us a bottle-neck and will need some attention.

**Comment by obastemur** [comment]
More context;

replaced the existing spin/sleep stuff with futex -> https://github.com/obastemur/ChakraCore/blob/fe23aeffe7513fcb0a1da996f11e144fc09c1dc4/pal/src/sync/CCSpinLock.cpp 

Doesn't make much difference on number of context switches. However I observe a consistent and repeating invalid call to futex that requires some attention.

**Comment by obastemur** [comment]
Surprisingly, https://github.com/obastemur/node-chakracore/blob/master/deps/chakrashim/src/v8objecttemplate.cc#L760-L805 is responsible from voluntary cont. switch difference between node-v8 and node-chakracore.

**Comment by MSLaguana** [comment]
Surely not directly? There aren't that many object templates with interceptors that we create; are you saying that we have a lot of calls to those few objects and those calls somehow cause us to context switch away?

**Comment by obastemur** [comment]
I don't have much details apart from the interesting fact I did share above. Commenting out that piece of code doesn't make much difference (no perf gain) yet.. ~360 vol. context switches are disappeared after the change.

I was trying to measure how significant the effect of the proxy. Only a single template and only for once the app uses that. Well, I wanted to cripple the whole proxy thing there by commenting the code out. However, after I did run the code again, I didn't receive any problem. However, cont. swit. numbers were interesting.
